include/deal.II/lac/trilinos_tpetra_precondition.h

```diff
@@ -0,0 +1,1311 @@
+#define dealii_trilinos_tpetra_precondition_h
+
+
+#include <deal.II/base/config.h>
+
+#include "deal.II/base/exceptions.h"
+#include "deal.II/base/memory_space.h"
+
+#include "deal.II/lac/vector.h"
+
+#include <Teuchos_BLAS_types.hpp>
+
+
+#ifdef DEAL_II_TRILINOS_WITH_TPETRA
+
+#  include <deal.II/base/subscriptor.h>
+
+#  include <deal.II/lac/la_parallel_vector.h>
+#  include <deal.II/lac/trilinos_tpetra_sparse_matrix.h>
+
+#  include <Teuchos_ConfigDefs.hpp>
+#  include <Teuchos_ParameterList.hpp>
+#  include <Teuchos_RCPDecl.hpp>
+#  include <Tpetra_Operator.hpp>
+
+
+DEAL_II_NAMESPACE_OPEN
+
+namespace LinearAlgebra
+{
+  namespace TpetraWrappers
+  {
+
+    /**
+     * The base class for all preconditioners based on Tpetra sparse matrices.
+     *
+     * @ingroup TpetraWrappers
+     * @ingroup Preconditioners
+     */
+    template <typename Number, typename MemorySpace = dealii::MemorySpace::Host>
+    class PreconditionBase : public Subscriptor
+    {
+    public:
+      /**
+       * @brief Standardized data struct to pipe additional flags to the
+       * preconditioner.
+       *
+       */
+      struct AdditionalData
+      {};
+
+      /**
+       * @brief Constructor. Does not do anything. The <tt>initialize</tt> function of
+       * the derived classes will have to create the preconditioner from a given
+       * sparse matrix.
+       *
+       */
+      PreconditionBase() = default;
+
+      /**
+       * @brief Desctructor.
+       * Destroys the preconditioner, leaving an object like just after having
+       * called the constructor.
+       */
+      void
+      clear();
+
+      /**
+       * @brief Apply the preconditioner.
+       *
+       * @param dst Input vector to apply the preconditioner to
+       * @param src Result vector
+       */
+      virtual void
+      vmult(Vector<Number, MemorySpace>       &dst,
+            const Vector<Number, MemorySpace> &src) const;
+
+      /**
+       * @brief Apply the transpose preconditioner
+       *
+       * @param dst Input vector to apply the preconditioner to
+       * @param src Result vector
+       */
+      virtual void
+      Tvmult(Vector<Number, MemorySpace>       &dst,
+             const Vector<Number, MemorySpace> &src) const;
+
+      /**
+       * @brief Apply the preconditioner
+       *
+       * @param dst Input vector to apply the preconditioner to
+       * @param src Result vector
+       */
+      virtual void
+      vmult(dealii::Vector<Number> &dst, dealii::Vector<Number> &src) const;
+
+      /**
+       * @brief Apply the transpose preconditioner
+       */
+      virtual void
+      Tvmult(dealii::Vector<Number> &dst, dealii::Vector<Number> &src) const;
+
+      /**
+       * @brief Access to underlying Trilinos data
+       *
+       * Calling this function from an uninitialized object will cause an
+       * exception.
+       */
+      Teuchos::RCP<TpetraTypes::LinearOperator<Number, MemorySpace>>
+      trilinos_operator() const;
```

Comment:
Since I might have missed the discussion elsewhere, could you please explain why you chose the `RCP` here, whereas we use https://github.com/dealii/dealii/blob/93dea5d7d15023c58bfedc3cabae9eb04c8ac92c/include/deal.II/lac/trilinos_tpetra_sparse_matrix.h#L996-L997 and https://github.com/dealii/dealii/blob/93dea5d7d15023c58bfedc3cabae9eb04c8ac92c/include/deal.II/lac/trilinos_tpetra_vector.h#L904-L905 which are the non-pointer references to the respective objects?

Reply:
A that was a mistake from starting with the Epetra Wrappers as a blueprint.
I will change it to `_operator()` returning the Type and `_rcp()` returning the rcp as in the vector and matrix classes.